### PR TITLE
Dismissing onboarding fix 

### PR DESCRIPTION
**Pull request by filip-solecki**
### Details

This PR is second attempt to solve [this issue](https://github.com/Expensify/App/issues/44401). Original PR was reverted due to issues listed below:
https://github.com/Expensify/App/issues/45157
https://github.com/Expensify/App/issues/45196 
https://github.com/Expensify/App/issues/44235
https://github.com/Expensify/App/issues/45284

### Fixed Issues

$ https://github.com/Expensify/App/issues/44401
PROPOSAL:

### Tests

**NOTE**
While testing on iOS web or Android web the app should not be running in the background on native devices

**CASE 1:**

1. Log in to the app as a new user
2. Onboarding modal should show
3. Replace `onboarding/purpose` with eg. `settings/profile` in the URL
4. You should still see onboarding Modal

**CASE 2:**

1. Log in to the app as existing user
2. Visit a few pages then log out
3. Log in as a new user
4. Try to go back using browser back button
5. You should see Onboarding Modal all the time you change url

- [x] Verify that no errors appear in the JS console

### QA Steps

While testing on iOS web or Android web the app should not be running in the background on native devices
### **Dismissing onboarding test steps**

**CASE 1:**
1. Log in to the app as a new user
2. Onboarding modal should show
3. Replace `onboarding/purpose` with eg. `settings/profile` in the URL
4. You should still see onboarding Modal

**CASE 2:**
1. Log in to the app as existing user
2. Visit a few pages then log out
3. Log in as a new user
4. Try to go back using browser back button
5. You should see Onboarding Modal all the time you change url

### **Previous regressions fix test steps**

**CASE 3**
1. Go to https://staging.new.expensify.com/
2. Log out
3. Navigate via public room link (my link https://staging.new.expensify.com/r/1353086947735283
)

**CASE 4**
1. Log out of NewDot
2. Navigate to https://staging.new.expensify.com/concierge
3. Sign-up with a new account
4. Finish the onboarding flow
5. You should see explanation message

**CASE 5**
1. Try to log in to the app with SSO
2. You should be logged in

**CASE 6**
1. Try to log in to the app with Magic link
2. You should be logged in

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: mWeb Chrome</summary>

[android.webm](https://github.com/user-attachments/assets/6a7739c1-1c10-49b1-889e-c533a5558723)

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/723b90c0-05cf-497e-824b-ba96709d7831

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/7ad9e52b-3ab9-4ae3-a3bc-5588b1148d66

</details>

<details>
<summary>SSO login issue</summary>

https://github.com/user-attachments/assets/8aa05f35-c68b-4992-8b65-edb5fa686b68

</details>

<details>
<summary>Magic link issue</summary>

https://github.com/user-attachments/assets/cb38845e-c8e5-486e-8f8d-ebf88e57bb6c

</details>

### CODE REVIEW

**Comment by filip-solecki** [comment]
cc @allgandalf as you were the reviewer of the original PR - basically I just added one check if the user is authenticated before we do some actions with navigation state.

**Comment by mountiny** [comment]
@filip-solecki can you please include all the issues that were identified as test/ qa in the PR?

**Comment by dangrous** [comment]
Just FYI - https://github.com/Expensify/App/issues/44600 is still open and happening on staging (calling it out because it says in the description here that the original PR was reverted... So it might not be related?

**Comment by mountiny** [comment]
It seems like that issue existed before this original pr no?

**Comment by filip-solecki** [comment]
> Just FYI - #44600 is still open and happening on staging. Calling it out because it says in the description here that the original PR was reverted... So it might not be related?
> 
> but if this PR fixes it, great!

sorry, my bad - I've made a mistake while copy pasting links - this issue is not fixed here

**Comment by filip-solecki** [comment]
@mountiny I've mentioned all issues I know are fixed here. There is one mentioned by @allgandalf and I don't know how to test it, maybe you have any idea @allgandalf ? 
Generally I think the PR is ready to review and the fixes are really small

**Comment by mountiny** [comment]
@mountiny @filip-solecki sorrg mostly ooo these days

I have triggered a build here

You can test it by intercepting the link to new dot with the shortlived token and change it for localhost or to the adhoc build link

**Review by allgandalf — APPROVED** [review]
I approve off this PR, LGTM, hopefully no issues arise this time 🙏

**Comment by filip-solecki** [comment]
Resolved 
cc @mountiny

**Review by mountiny — APPROVED** [review]
Thanks for the patience!

**Comment by mvtglobally** [comment]
Onboarding still can be skipped by navigating to page with RHP. This PR is failing with the original issue https://github.com/Expensify/App/issues/44401

https://github.com/user-attachments/assets/9623a096-44e5-466d-9f40-1345a1d1530f

**Comment by filip-solecki** [comment]
@adamgrzybowski is taking a look

**Comment by WojtekBoman** [comment]
> Onboarding still can be skipped by navigating to page with RHP. This PR is failing with the original issue #44401
> 
>  45327.web.mp4

Hi! I'll take a look at it :)

**Comment by WojtekBoman** [comment]
> Onboarding still can be skipped by navigating to page with RHP. This PR is failing with the original issue #44401
> 
>  45327.web.mp4

Could you recheck it? I tried to reproduce that and it seems to work fine :) I tested it on a local build and on staging, in both cases it behaves correctly. 


https://github.com/user-attachments/assets/ecbdf9ac-4503-4fad-b893-01a67ddc9d52

File: src/libs/Navigation/AppNavigator/createCustomStackNavigator/CustomRouter.ts [hunk 1]
```diff
@@ -97,6 +100,23 @@ function compareAndAdaptState(state: StackNavigationState<RootStackParamList>) {
...
+    }
+    const currentFocusedRoute = findFocusedRoute(state);
+    const targetFocusedRoute = findFocusedRoute(action?.payload);
+
+    // We want to prevent the user from navigating back to a non-onboarding screen if they are currently on an onboarding screen
+    if (isOnboardingFlowName(currentFocusedRoute?.name) && !isOnboardingFlowName(targetFocusedRoute?.name)) {
+        Welcome.setOnboardingErrorMessage(Localize.translateLocal('onboarding.purpose.errorBackButton'));
+        // We reset the URL as the browser sets it in a way that doesn't match the navigation state
+        // eslint-disable-next-line no-restricted-globals
+        history.replaceState({}, '', getPathFromState(state, linkingConfig.config));
```
Comment (s77rt) [inline src/libs/Navigation/AppNavigator/createCustomStackNavigator/CustomRouter.ts:115]: `history` does not exist on native. I am not sure if a real user could land into this code but I got a crash while testing another PR

Screenshot 2024-10-09 at 8 07 01 PM

Comment (adamgrzybowski) [reply src/libs/Navigation/AppNavigator/createCustomStackNavigator/CustomRouter.ts:115]: Yeah, we definetly should make this part platform-specific.

Comment (adamgrzybowski) [reply src/libs/Navigation/AppNavigator/createCustomStackNavigator/CustomRouter.ts:115]: Somebody from SWM can take care of this issue if you want

Comment (mountiny) [reply src/libs/Navigation/AppNavigator/createCustomStackNavigator/CustomRouter.ts:115]: https://github.com/Expensify/App/issues/50713

File: src/libs/Navigation/NavigationRoot.tsx [hunk 2]
```diff
@@ -76,26 +80,44 @@ function NavigationRoot({authenticated, lastVisitedPath, initialUrl, onReady}: N
...
-            const path = initialUrl ? getPathFromURL(initialUrl) : null;
-
-            // For non-nullable paths we don't want to set initial state
-            if (path) {
-                return;
-            }
+    const initialState = useMemo(() => {
+        if (!user || user.isFromPublicDomain) {
+            return;
+        }
```
Comment (s77rt) [inline src/libs/Navigation/NavigationRoot.tsx:92]: This return is too early and prevented public domain users from using the `lastVisitedPath` feature.

**Repro steps:**
1. Go to settings
2. Open https://dev.new.expensify.com:8082/
3. Verify that you land on settings

This only works with private domain users. cc @mountiny 


https://github.com/user-attachments/assets/fd4e6d69-4a42-492e-bed3-e88144e6974c

**Pull request merged**